### ISSUE TITLE
Array out of bounds in GLSL caused by incorrect SPIR-V

### ISSUE THREAD

**Issue by redorav** [KhronosGroup/SPIRV-Tools#1203]
I reported a bug over at the SPIRV-Cross project about an array out of bounds issue where a variable named 645 is being used as a literal to access an array (so the code in glsl ends up as Array[645] instead of Array[_645]. The bug is this https://github.com/KhronosGroup/SPIRV-Cross/issues/397

This seems to be a very recent change as I'm on 5eafc00 commit eb0c73dad6102fc0d4f03c62fe910348bae43a11

which is only 5 days ago. I don't have a simple repro but the SPIRV-Cross bug contains the faulty SPIR-V.

**Comment by alan-baker** [comment]
@redorav can you share the input to spirv-opt?

**Comment by redorav** [comment]
I've had to redo the repro and I lost the original one, but here's another shader that shows the same issue.

[ArrayOutOfBounds_spv.txt](https://github.com/KhronosGroup/SPIRV-Tools/files/1635849/ArrayOutOfBounds_spv.txt)
[ArrayOutOfBounds_optimized_spv.txt](https://github.com/KhronosGroup/SPIRV-Tools/files/1635850/ArrayOutOfBounds_optimized_spv.txt)

The resulting glsl shows the following code:

```glsl
for (int _484 = 0; _484 < 5; _484++)
{
    _486[_484] = clamp(In_uvSet01.xy + _23.g_RingEffectPS.ringOffsets[484].xy, _23.g_RingEffectPS.uvClamp.xy, _23.g_RingEffectPS.uvClamp.zw);
}
```

As you can see, ringOffsets is using 484 directly, instead of _484 as it's meant to be using

**Comment by redorav** [comment]
I'm using a custom recipe, this is the code. I try to base these on the performance passes, but I've had to sometimes disable some that cause bugs which is why I keep a custom one. Perhaps I'm doing something wrong?

```hlsl
optimizer.RegisterPass(spvtools::CreateRemoveDuplicatesPass());
optimizer.RegisterPass(spvtools::CreateMergeReturnPass());
optimizer.RegisterPass(spvtools::CreateInlineExhaustivePass());
optimizer.RegisterPass(spvtools::CreateEliminateDeadFunctionsPass());
optimizer.RegisterPass(spvtools::CreateScalarReplacementPass());
optimizer.RegisterPass(spvtools::CreateLocalAccessChainConvertPass());
optimizer.RegisterPass(spvtools::CreateLocalSingleBlockLoadStoreElimPass());
optimizer.RegisterPass(spvtools::CreateLocalSingleStoreElimPass());
optimizer.RegisterPass(spvtools::CreateInsertExtractElimPass());
optimizer.RegisterPass(spvtools::CreateCCPPass());
optimizer.RegisterPass(spvtools::CreateAggressiveDCEPass());
optimizer.RegisterPass(spvtools::CreateDeadBranchElimPass());
optimizer.RegisterPass(spvtools::CreateBlockMergePass());
//optimizer.RegisterPass(spvtools::CreateLocalMultiStoreElimPass());
optimizer.RegisterPass(spvtools::CreateInsertExtractElimPass());
optimizer.RegisterPass(spvtools::CreateRedundancyEliminationPass());
optimizer.RegisterPass(spvtools::CreateLocalRedundancyEliminationPass());
optimizer.RegisterPass(spvtools::CreateCFGCleanupPass());
optimizer.RegisterPass(spvtools::CreateCommonUniformElimPass());
optimizer.RegisterPass(spvtools::CreateDeadVariableEliminationPass());
optimizer.RegisterPass(spvtools::CreateEliminateDeadConstantPass());
optimizer.RegisterPass(spvtools::CreateStrengthReductionPass());
optimizer.RegisterPass(spvtools::CreateUnifyConstantPass());
```

**Comment by alan-baker** [comment]
I've successfully reproduced and root-caused this issue. Introduced by 6587d3f8a31b636c0d75b32d5ae65d338581c200. Bug is in CommonUniformElimPass::IsConstantIndexAccessChain. Missed a variable update when converting to WhileEach.

**Comment by alan-baker** [comment]
Fixed by #1204
